src/MongoDB.Driver.Core/Core/Authentication/SaslPrepHelper.cs

```diff
@@ -0,0 +1,805 @@
+        private static bool ChangeDisplayProperties(int codepoint)
+        {
+            return codepoint == 0x0340
+                || codepoint == 0x0341
+                || codepoint == 0x200E
+                || codepoint == 0x200F
+                || codepoint == 0x202A
+                || codepoint == 0x202B
+                || codepoint == 0x202C
+                || codepoint == 0x202D
+                || codepoint == 0x202E
+                || codepoint == 0x206A
+                || codepoint == 0x206B
+                || codepoint == 0x206C
+                || codepoint == 0x206D
+                || codepoint == 0x206E
+                || codepoint == 0x206F;
+        }
+
+        /// <summary>
+        /// Returns the number of characters required to represent a specified Unicode character.
+        /// </summary>
+        /// <param name="codepoint">The Unicode character's codepoint.</param>
+        /// <returns>Number of characters required to represent a specified Unicode character.</returns>
+        private static int CharCount(int codepoint)
+        {
+            return codepoint >= 0x10000 ? 2 : 1;
+        }
+        
+        /// <summary>
+        /// Return true if the given <paramref name="codepoint"/> is inappropriate for canonical representation
+        /// characters as defined by <a href="https://tools.ietf.org/html/rfc3454#appendix-C.7">RFC 3454, Appendix C.7</a>. 
+        /// </summary>
+        /// <param name="codepoint">The Unicode character's codepoint.</param>
+        /// <returns></returns>
+        private static bool InappropriateForCanonical(int codepoint)
+        {
+            return 0x2FF0 <= codepoint && codepoint <= 0x2FFB;
+        }
+
+        /// <summary>
+        /// Return true if the given <paramref name="codepoint"/> is inappropriate for plain text characters as defined
+        /// by <a href="https://tools.ietf.org/html/rfc3454#appendix-C.6">RFC 3454, Appendix C.6</a>.
+        /// </summary>
+        /// <param name="codepoint">The Unicode character's codepoint.</param>
+        /// <returns></returns>
+        private static bool InappropriateForPlainText(int codepoint)
+        {
+            return codepoint == 0xFFF9
+                || codepoint == 0xFFFA
+                || codepoint == 0xFFFB
+                || codepoint == 0xFFFC
+                || codepoint == 0xFFFD;
+        }
+
+        /// <summary>
+        /// Returns whether or not a Unicode character represented by a codepoint is defined in Unicode.
+        /// See <a href="https://tools.ietf.org/html/rfc5137">RFC 5137</a> for more details.
+        /// </summary>
+        /// <param name="codepoint">The Unicode character's codepoint.</param>
+        /// <returns>Whether or not the Unicode character represnted by codepoint is defined in Unicode.</returns>
+        private static bool IsDefined(int codepoint)
+        {   
+            return IsSurrogateCodepoint(codepoint) || 
```

Comment:
See offline discussion for proof of correctness.

Reply:
Can you repeat that here please?  I'm not clear on why the first condition is needed.

Reply:
Nice.  And nice use of jshell!

Reply:
Thanks! 😺